action/protocol/staking/handlers.go

```diff
@@ -425,7 +429,28 @@ func (p *Protocol) handleRestake(ctx context.Context, act *action.Restake, csm C
 
 	prevWeightedVotes := p.calculateVoteWeight(bucket, csm.ContainsSelfStakingBucket(act.BucketIndex()))
 	// update bucket
-	bucket.StakedDuration = time.Duration(act.Duration()) * 24 * time.Hour
+	actDuration := time.Duration(act.Duration()) * 24 * time.Hour
+	if bucket.StakedDuration.Hours() > actDuration.Hours() {
+		// in case of reducing the duration
+		if bucket.AutoStake {
+			// if auto-stake on, user can't reduce duration
+			err := errors.New("AutoStake should be disabled first in order to reduce duration")
+			log.L().Debug("Error when restaking bucket", zap.Error(err))
+			return p.settleAction(ctx, csm, uint64(iotextypes.ReceiptStatus_ErrInvalidBucketType), gasFee)
+		}
+		if !bucket.AutoStake && blkCtx.BlockTimeStamp.Before(bucket.StakeStartTime.Add(bucket.StakedDuration)) {
+			// if auto-stake off and maturity is not enough
+			err := fmt.Errorf("bucket is not ready to be able to reduce duration, current time %s, required time %s",
+				blkCtx.BlockTimeStamp, bucket.StakeStartTime.Add(bucket.StakedDuration))
+			log.L().Debug("Error when restaking bucket", zap.Error(err))
+			return p.settleAction(ctx, csm, uint64(iotextypes.ReceiptStatus_ErrReduceDurationBeforeMaturity), gasFee)
+		}
+	}
+	bucket.StakedDuration = actDuration
+	if bucket.AutoStake && !act.AutoStake() {
```

Comment:
Shouldn't we update the stakeStartTime no matter what?

Reply:
i have the same question, we should, right?

Reply:
It can be 3 cases, 
1. not change autostake (only changing duration) 
2. switch-on 
3. switch-off 
All of those case, we want to restart the staking time elapse? 

such as duration was 7 days, but he changed to 30 days at certain point. 
then he have to wait another 30 days is that right?

Reply:
make sense, address on latest commit